source/framework/core/src/TRestDataSet.cxx

```diff
@@ -0,0 +1,543 @@
+/// (or observables) that have been defined by the user.
+///
+void TRestDataSet::Initialize() {
+    SetSectionName(this->ClassName());
+
+    if (fTree != nullptr) {
+        RESTWarning << "Tree has already been loaded. Skipping TRestDataSet::LoadDataFrame ... " << RESTendl;
+        return;
+    }
+
+    if (fFileSelection.empty()) FileSelection();
+
+    // We are not ready yet
+    if (fFileSelection.empty()) return;
+
+    ///// Disentangling process observables --> producing finalList
+    TRestRun* run = new TRestRun(fFileSelection[0]);
+    std::vector<std::string> finalList;
+    finalList.push_back("runOrigin");
+    finalList.push_back("eventID");
+    finalList.push_back("timeStamp");
+
+    for (const auto& obs : fObservablesList) finalList.push_back(obs);
+
+    std::vector<std::string> obsNames = run->GetAnalysisTree()->GetObservableNames();
+    for (const auto& name : obsNames) {
+        for (const auto& pcs : fProcessObservablesList) {
+            if (name.find(pcs) == 0) finalList.push_back(name);
+        }
+    }
+    delete run;
+    ///////
+
+    ROOT::EnableImplicitMT();
+
+    fDataSet = ROOT::RDataFrame("AnalysisTree", fFileSelection);
+
+    std::string user = getenv("USER");
+    std::string fOutName = "/tmp/rest_output_" + user + ".root";
+    fDataSet.Snapshot("AnalysisTree", fOutName, finalList);
+
+    fDataSet = ROOT::RDataFrame("AnalysisTree", fOutName);
+
+    TFile* f = TFile::Open(fOutName.c_str());
+    fTree = (TTree*)f->Get("AnalysisTree");
+
+    int cont = 0;
+    std::string obsListStr;
+    for (const auto& l : finalList) {
+        if (cont > 0) obsListStr += ":";
+        obsListStr += l;
+        cont++;
+    }
+
+    // We do this so that later we can recover the values using TTree::GetVal
+    fTree->Draw((TString)obsListStr, "", "goff");
+
+    std::cout << " - Dataset initialized!" << std::endl;
+}
+
+///////////////////////////////////////////////
+/// \brief Function to determine the filenames that satisfy the dataset conditions
+///
+std::vector<std::string> TRestDataSet::FileSelection() {
+    fFileSelection.clear();
+
+    std::time_t time_stamp_start = REST_StringHelper::StringToTimeStamp(fStartTime);
+    std::time_t time_stamp_end = REST_StringHelper::StringToTimeStamp(fEndTime);
+
+    if (!time_stamp_end || !time_stamp_start) {
+        RESTError << "TRestDataSet::FileSelect. Start or end dates not properly formed. Please, check "
+                     "REST_StringHelper::StringToTimeStamp documentation for valid formats"
+                  << RESTendl;
+        return fFileSelection;
+    }
+
+    std::vector<std::string> fileNames = TRestTools::GetFilesMatchingPattern(fFilePattern);
+
+    if (!fileNames.empty()) {
+        RESTInfo << " - Starting to select files matching pattern. Total files : " << fileNames.size()
+                 << RESTendl;
+    }
+
+    fTotalDuration = 0;
+    for (const auto& file : fileNames) {
+        TRestRun* run = new TRestRun(file);
```

Comment:
Again `new` and `delete`, please use `TRestRun run(file);`

Reply:
For me when I do `new` and `delete` it is clear the instance is created and destroyed. Don't see why this is a problem here.

Reply:
I just want to make sure the instance is destroyed.

Reply:
We should minimize the use of `new` and there is no reason to use it here
https://stackoverflow.com/questions/6500313/why-should-c-programmers-minimize-use-of-new